spinn_front_end_common/utility_models/data_speed_up_packet_gatherer_machine_vertex.py

```diff
@@ -0,0 +1,754 @@
+        AbstractProvidesIncomingPartitionConstraints.__init__(self)
+        AbstractProvidesLocalProvenanceData.__init__(self)
+
+        # data holders for the output, and seq nums
+        self._view = None
+        self._max_seq_num = None
+        self._output = None
+
+        # create socket
+        if connection is None:
+            self._connection = UDPConnection(local_host=None)
+        else:
+            self._connection = connection
+
+        # local provenance storage
+        self._provenance_data_items = defaultdict(list)
+
+    @property
+    @overrides(MachineVertex.resources_required)
+    def resources_required(self):
+        return self.resources_required_for_connection(self._connection)
+
+    @staticmethod
+    def resources_required_for_connection(connection):
+        """ used to generate resources required given a UDP connection
+        Exists for giving the app vertex access to the data without clones.
+
+        :param connection: UDP connection for transmissions
+        :return: resource container
+        """
+        return ResourceContainer(
+            sdram=SDRAMResource(
+                constants.SYSTEM_BYTES_REQUIREMENT +
+                DataSpeedUpPacketGatherMachineVertex.CONFIG_SIZE),
+            iptags=[IPtagResource(
+                port=connection.local_port, strip_sdp=True,
+                ip_address="localhost")])
+
+    @inject_items({"machine_graph": "MemoryMachineGraph"})
+    @overrides(AbstractProvidesIncomingPartitionConstraints.
+               get_incoming_partition_constraints,
+               additional_arguments={"machine_graph"})
+    def get_incoming_partition_constraints(self, partition, machine_graph):
+        if partition.identifier != \
+                constants.PARTITION_ID_FOR_MULTICAST_DATA_SPEED_UP:
+            raise Exception("do not recognise this partition identifier")
+
+        vertex_partition = list()
+        for incoming_edge in machine_graph.get_edges_ending_at_vertex(self):
+            partition = \
+                machine_graph.get_outgoing_edge_partition_starting_at_vertex(
+                    incoming_edge.pre_vertex,
+                    constants.PARTITION_ID_FOR_MULTICAST_DATA_SPEED_UP)
+            vertex_partition.append(partition)
+        return self.static_get_incoming_partition_constraints(
+            partition, vertex_partition)
+
+    @staticmethod
+    def static_get_incoming_partition_constraints(partition, vertex_partition):
+        constraints = list()
+        # if partition.traffic_type == EdgeTrafficType.MULTICAST:
+        #    constraints.append(
+        #        FixedKeyAndMaskConstraint([
+        #            BaseKeyAndMask(
+        #                DataSpeedUpPacketGatherMachineVertex.BASE_KEY,
+        #                DataSpeedUpPacketGatherMachineVertex.BASE_MASK
+        #            )]))
+        #    constraints.append(ShareKeyConstraint(vertex_partition))
+        return constraints
```

Comment:
TODO: why all the commented out code?

Reply:
this is to support mc performance when/if its needed. can delete if ya want